cmd/agent/app/server.go

```diff
@@ -0,0 +1,137 @@
+package app
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/http"
+	"net/http/pprof"
+	"runtime"
+
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/spf13/cobra"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"k8s.io/klog/v2"
+
+	"sigs.k8s.io/apiserver-network-proxy/cmd/agent/app/options"
+	"sigs.k8s.io/apiserver-network-proxy/pkg/util"
+)
+
+func NewAgentCommand(a *Agent, o *options.GrpcProxyAgentOptions) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:  "agent",
+		Long: `A gRPC agent, Connects to the proxy and then allows traffic to be forwarded to it.`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return a.run(o)
+		},
+	}
+
+	return cmd
+}
+
+type Agent struct {
+}
+
+func (a *Agent) run(o *options.GrpcProxyAgentOptions) error {
+	o.Print()
+	if err := o.Validate(); err != nil {
+		return fmt.Errorf("failed to validate agent options with %v", err)
+	}
+
+	stopCh := make(chan struct{})
+	if err := a.runProxyConnection(o, stopCh); err != nil {
+		return fmt.Errorf("failed to run proxy connection with %v", err)
+	}
+
+	if err := a.runHealthServer(o); err != nil {
+		return fmt.Errorf("failed to run health server with %v", err)
+	}
+
+	if err := a.runAdminServer(o); err != nil {
+		return fmt.Errorf("failed to run admin server with %v", err)
+	}
+
+	<-stopCh
+
+	return nil
+}
+
+func (a *Agent) runProxyConnection(o *options.GrpcProxyAgentOptions, stopCh <-chan struct{}) error {
+	var tlsConfig *tls.Config
+	var err error
+	if tlsConfig, err = util.GetClientTLSConfig(o.CaCert, o.AgentCert, o.AgentKey, o.ProxyServerHost, o.AlpnProtos); err != nil {
+		return err
+	}
+	dialOption := grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
+	cc := o.ClientSetConfig(dialOption)
+	cs := cc.NewAgentClientSet(stopCh)
+	cs.Serve()
+
+	return nil
+}
+
+func (a *Agent) runHealthServer(o *options.GrpcProxyAgentOptions) error {
+	livenessHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "ok")
+	})
+	readinessHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "ok")
+	})
+
+	muxHandler := http.NewServeMux()
+	muxHandler.Handle("/metrics", promhttp.Handler())
+	muxHandler.HandleFunc("/healthz", livenessHandler)
+	muxHandler.HandleFunc("/ready", readinessHandler)
```

Comment:
I realize this was a pre-existing bug but this should be ```/readyz```.

Reply:
I found that it is /ready in master branch. Maybe we can fix it in other PRs. This PR only refine main package.